rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+
+If all 3 of these criteria are true for `result.url`:
+- The domain of the URL  matches the domain of the `publicUrl`
+- The pathname of the URL starts with the current basePath (eg. /mybasepath/s/my-space)
+- The pathname segment after the basePath matches an application route (eg. /app/<id>/)
+
+Then: match the pathname segment to the corresponding application and do the SPA navigation to that application using 
+the remaining pathname segment
+
+Otherwise: do a full page navigation (`window.location.assign()`)
+
+### searching from the server side
+
+When calling `GlobalSearchServiceStart.find` from the server-side service:
+
+- the service will call `find` on each server-side registered result provider and collect the resulting result observables
+
+- then, the service will merge every result observable and trigger the next step on every emission until either
+    - A predefined timeout duration is reached
+    - All result observables are completed
+   
+- on every emission of the merged observable, the results will be aggregated to the existing list and sorted following 
+the logic defined in the [results aggregation](#results-aggregation) section
+
+A very naive implementation of this behavior would be:
+
+```ts
+search(
+  term: string,
+  options: GlobalSearchOptions,
+  request: KibanaRequest
+): Observable<GlobalSearchResponse> {
+  const fromProviders$ = this.providers.map(p =>
+    p.find(term, options, contextFromRequest(request))
+  );
+  const results: GlobalSearchResult[] = [];
+  return merge([...fromProviders$]).pipe(
+    takeUntil(timeout$),
+    map(newResults => {
+      results.push(...newResults);
+      return order(results);
+    }),
+  );
+}
+```
+
+### searching from the client side
+
+When calling `GlobalSearchServiceStart.find` from the public-side service:
+
+- The service will call:
+  - the server-side API via an http call to fetch results from the server-side result providers
+  - `find` on each client-side registered result provider and collect the resulting observables
+
+- Then, the service will merge every result observable and trigger the next step on every emission until either
+  - A predefined timeout duration is reached
+  - All result observables are completed
+   
+- On every emission of the merged observable, the results will be aggregated to the existing list and sorted following 
+the logic defined in the [results aggregation](#results-aggregation) section
+
+A very naive implementation of this behavior would be:
+
+```
+search(
+  term: string,
+  options: GlobalSearchOptions,
+): Observable<GlobalSearchResponse> {
+  const fromProviders$ = this.providers.map(p =>
+    p.find(term, options)
+  );
+  const fromServer$ = of(this.fetchServerResults(term, options))
+
+  const results: GlobalSearchResult[] = [];
+  return merge([...fromProviders$, fromServer$]).pipe(
+    takeUntil(timeout$),
+    map(newResults => {
+      results.push(...newResults);
+      return order(results);
+    }),
+  );
+}
```

Comment:
```ts
this.fetchServerResults = plugins.bfetch.batchedFunction({
  url: '/global_search',
});
```

Reply:
This seems to be exactly what we need.

However as `bfetch` is a plugin, this means it's not usable from `core`. I guess that's another solid argument to provide the `GS` API as a plugin instead of a core service.

@joshdover wdyt?

Reply:
Does `bfetch` help us here? We're going to have multiple XHRs from the client-side, one of which we know will be for a Kibana server API; however, there will be other requests to Cloud specific APIs. In terms of HTTP requests, how does `bfetch` work? We'd have to enforce that the Cloud specific APIs abide by this contract.

Reply:
I'm also unclear on how bfetch applies here. From my understanding it allows you to make multiple requests to the same backend function and avoid opening new connections for each call.

In this case we're making a single request to the backend which may have multiple _different_ providers that need their results streamed back as they arrive. We don't need to make many search requests at once and get all of their results. All we care about are the results from the last search request, which may contain results from several different sources.

Maybe I'm missing something here?

Reply:
> We're going to have multiple XHRs from the client-side, one of which we know will be for a Kibana server API; however, there will be other requests to Cloud specific APIs

Even if the 'cloud' result provider is not designed yet, it will very likely be on the server-side, meaning that this assertion is wrong: the client will only perform ONE request to the server-side GS internal API. kibana to cloud calls will be performed from the server.

Reply:
> I'm also unclear on how bfetch applies here. From my understanding it allows you to make multiple requests to the same backend function and avoid opening new connections for each call.

> In this case we're making a single request to the backend which may have multiple different providers that need their results streamed back as they arrive. We don't need to make many search requests at once and get all of their results

Maybe I also misunderstood the actual usage of `bfetch`. In the example implementation at the start of this discussion, there is only ONE http call: 

```ts
fromServer$ = of(this.fetchServerResults(term, options))
```

The other providers are the client-side ones, so how they choose to fetch/emit their result is an implementation detail outside of the perimeter of our implementation.

See this remark:

```
Due to the complexity of the process, the initial implementation will not be streaming results from the server,
meaning that all results from server-side registered providers will all be fetched at the same time (via a 'classic' 
http call to the GS endpoint)
```

So the only question here is: is `bfetch` capable to open and keep opened a single request to our GS backend to let it streams the results from the backend result providers observables, or should we just stick with a classic http call returning them all at one (once they are all emitted from the backend providers)

@streamich ^

Reply:
Also second point is: let be realistic, our initial result providers (SO and app) will not really use this observable API at all, as:
- the app client-side provider will just grep in the registered apps label and return all it's results in a single emission
- the SO server-side provider will likely only perform a single `find` request (at least for typeahead) and return all the results in a single emission too.

So maybe we are just over-engineering this thing, and awaiting all the results from the server-side before sending them to the client is just good enough?

Reply:
`bfetch` does two things:

1. It can batch multiple requests into one and send them to Kibana server in a single *request*.
2. On the *response* side, it streams data as soon as it is available; you don't need to build up the whole request and send it all at once.

Reply:
My understanding is that server-side API is Observable-based, so if you want to stream them to the client as soon as they are available, `bfetch` could do that.